x-pack/legacy/plugins/apm/server/routes/settings.ts

```diff
@@ -80,20 +81,9 @@ export const listAgentConfigurationEnvironmentsRoute = createRoute(() => ({
 }));
 
 // create configuration
-const transactionSampleRateRt = new t.Type<number, number, unknown>(
-  'TransactionSampleRate',
-  (u): u is number => typeof u === 'number',
-  (u, c) => {
-    return Number(u) >= 0 && Number(u) <= 1
-      ? t.success(Number(u))
-      : t.failure(u, c);
-  },
-  a => Number(a.toPrecision(3))
-);
-
 const agentPayloadRt = t.type({
   settings: t.type({
-    transaction_sample_rate: transactionSampleRateRt
+    transaction_sample_rate: createRangeType(0, 1, 3)
```

Comment:
Hmm... maybe I'm just old-school but I find this a lot simpler:
```js
settings: t.type({
  transaction_sample_rate: t.number
})
```

I know this lacks the the runtime check, but we can handle that in the handler somehow.
```js
const schema = Joi.number()
  .min(0)
  .max(1)
  .precision(3)
  .required()
  .options({ convert: false })

const value = 1;
Joi.validate(value, schema);
```

Is that terrible? I just find the io-ts dsl for this kind of stuff so bad.

Reply:
Either way, feel free to keep it as-is.

Reply:
Actually, you could extract the validation logic used on the client, and re-use that on the server?

Reply:
Something like:
```js
function validateTransactionSampleRate(sampleRate) {
  const hasCorrectDecimals = Number.isInteger(sampleRate * 1000);
  return sampleRate >= 0 && sampleRate <= 1 && hasCorrectDecimals;
}
```

Reply:
Are you worried about `createRangeType()`, or the implementation of the range type factory function? I think that if we can avoid custom error handling, we should (which is what we would have to add if we inlined Joi or used a plain function). We could also pass `options.validators`, but that would be kind of gross too (I'm thinking we either use Joi, or we don't).

Reply:
btw, I don't think the io-ts DSL for creating new types is that bad, they're pretty composable as well which is nice. Here's an example of piping a JSON type into a more specific type that parses `filterNames` as JSON and then checks the value of the resulting array: https://github.com/elastic/kibana/pull/42961/files#diff-dcf737a7d566dbcc41bb9bac8ca54ff9R40

Reply:
Sorry I wrote a lot of comments above and didn't make myself clear.
And I'm fine with moving forward with the io-ts type if we can use it on the client and on the server. If we can't I think we should extract the custom validation logic from the client, and use it both on server and client.

Reply:
Ok! Should I use the `io-ts` type for the agent configuration payload on the client as well in this PR, or open up a separate PR for that one?

Reply:
I'd like to see how it would look before we merge this. I'm not convinced that the simples solution is to use io-ts on server and client (compared to sharing custom logic), but I probably need to see it first.

I don't think the signature for this is great: `createRangeType(0, 1, 3)`. Looking at it I have no idea what the arguments are.  Probably it would better just to have `transactionSampleRate()` and hardcode the values inside it.